File: backend/app/schemas/route.py
from typing import Any, Literal

from pydantic import BaseModel, Field

from app.services.amap.schemas import AmapRouteMode


class RoutePoi(BaseModel):
    id: str
    name: str
    longitude: float
    latitude: float


class RouteChainRequest(BaseModel):
    mode: AmapRouteMode
    pois: list[RoutePoi]


class RouteSegmentSummary(BaseModel):
    segment_index: int
    from_poi_id: str
    from_poi_name: str
    to_poi_id: str
    to_poi_name: str
    distance_m: float
    duration_s: float


class RouteStepFeatureProperties(BaseModel):
    segment_index: int
    step_index: int
    from_poi_id: str
    from_poi_name: str
    to_poi_id: str
    to_poi_name: str
    instruction: str | None = None
    road_name: str | None = None
    distance_m: float
    duration_s: float | None = None


class GeoJSONLineString(BaseModel):
    type: Literal["LineString"] = "LineString"
    coordinates: list[list[float]]


class GeoJSONFeature(BaseModel):
    type: Literal["Feature"] = "Feature"
    properties: RouteStepFeatureProperties
    geometry: GeoJSONLineString


class GeoJSONFeatureCollection(BaseModel):
    type: Literal["FeatureCollection"] = "FeatureCollection"
    features: list[GeoJSONFeature] = Field(default_factory=list)


class RouteChainResponse(BaseModel):
    mode: AmapRouteMode
    ordered_pois: list[RoutePoi]
    total_distance_m: float
    total_duration_s: float
    segments: list[RouteSegmentSummary]
    geojson: GeoJSONFeatureCollection


ErrorDetail = dict[str, Any]

File: frontend/src/types/route.ts
export type RouteMode = "walking" | "driving"

export interface RoutePoi {
  id: string
  name: string
  longitude: number
  latitude: number
  category?: string | null
  cover_image?: string | null
}

export interface RouteChainRequest {
  mode: RouteMode
  pois?: RoutePoi[]
  poi_ids?: string[]
}

export interface RouteSegmentSummary {
  segment_index: number
  from_poi_id: string
  from_poi_name: string
  to_poi_id: string
  to_poi_name: string
  distance_m: number
  duration_s: number
}

export interface RouteStepFeatureProperties {
  segment_index: number
  step_index: number
  from_poi_id: string
  from_poi_name: string
  to_poi_id: string
  to_poi_name: string
  instruction?: string | null
  road_name?: string | null
  distance_m: number
  duration_s?: number | null
}

export interface GeoJSONLineString {
  type: "LineString"
  coordinates: [number, number][]
}

export interface GeoJSONFeature {
  type: "Feature"
  properties: RouteStepFeatureProperties
  geometry: GeoJSONLineString
}

export interface GeoJSONFeatureCollection {
  type: "FeatureCollection"
  features: GeoJSONFeature[]
}

export interface RouteChainResponse {
  mode: RouteMode
  ordered_pois: RoutePoi[]
  total_distance_m: number
  total_duration_s: number
  segments: RouteSegmentSummary[]
  geojson: GeoJSONFeatureCollection
}

export interface AmapRouteRequest {
  mode: RouteMode
  poi_ids: string[]
  source: "ugc_instant_route" | "manual_route"
  pool_id?: string
  free_text?: string
  date?: string
  time_window?: {
    start: string
    end: string
  }
}
